src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,381 @@
+ * for efficient and easy use.
+ */
+public final class CaptureStateUtil {
+
+    private static final StructLayout CAPTURE_LAYOUT = Linker.Option.captureStateLayout();
+    private static final BufferStack POOL = BufferStack.of(CAPTURE_LAYOUT);
+
+    // Do not use a lambda in order to allow early use in the init sequence
+    private static final Function<BasicKey, MethodHandle> UNDERLYING_MAKE_BASIC_HANDLE = new Function<>() {
+        @Override
+        public MethodHandle apply(BasicKey basicKey) {
+            return makeBasicHandle(basicKey);
+        }
+    };
+
+    // The `BASIC_HANDLE_CACHE` contains the common "basic handles" that can be reused for
+    // all adapted method handles. Keeping as much as possible reusable reduces the number
+    // of combinators needed to form an adapted method handle.
+    // The function is lazily computed.
+    //
+    private static final Function<BasicKey, MethodHandle> BASIC_HANDLE_CACHE;
+
+    static {
+        final Set<BasicKey> inputs = new HashSet<>();
+        // The Cartesian product : (int.class, long.class) x ("errno", ...)
+        // Do not use Streams in order to enable "early" use in the init sequence.
+        for (Class<?> c : new Class<?>[]{int.class, long.class}) {
+            for (MemoryLayout layout : CAPTURE_LAYOUT.memberLayouts()) {
+                inputs.add(new BasicKey(c, layout.name().orElseThrow()));
+            }
+        }
+        BASIC_HANDLE_CACHE = StableValue.function(inputs, UNDERLYING_MAKE_BASIC_HANDLE);
+    }
+
+    // A key that holds both the `returnType` and the `stateName` needed to look up a
+    // specific "basic handle" in the `BASIC_HANDLE_CACHE`.
+    //   returnType in {int.class | long.class}
+    //   stateName can be anything non-null but should be in {"GetLastError" | "WSAGetLastError" | "errno"}
+    private record BasicKey(Class<?> returnType, String stateName) {
+
+        BasicKey(MethodHandle target, String stateName) {
+            this(returnType(target), Objects.requireNonNull(stateName));
+        }
+
+        static Class<?> returnType(MethodHandle target) {
+            // Implicit null check
+            final MethodType type = target.type();
+            final Class<?> returnType = type.returnType();
+
+            if (!(returnType.equals(int.class) || returnType.equals(long.class))) {
+                throw illegalArgDoesNot(target, "return an int or a long");
+            }
+            if (type.parameterCount() == 0 || type.parameterType(0) != MemorySegment.class) {
+                throw illegalArgDoesNot(target, "have a MemorySegment as the first parameter");
+            }
+            return returnType;
+        }
+
+        private static IllegalArgumentException illegalArgDoesNot(MethodHandle target, String info) {
+            return new IllegalArgumentException("The provided target " + target
+                    + " does not " + info);
+        }
+
+        @Override
+        public int hashCode() {
```

Comment:
isn't the default toString/hashCode/equals for records good enough?

Reply:
Added for bootstrapping but I think this could be addressed in another way (e.g. helloClass once this class is used).